idb/postgres/postgres.go

```diff
@@ -871,52 +871,97 @@ func (db *IndexerDb) CommitRoundAccounting(updates idb.RoundUpdates, round, rewa
 		}
 		defer uta.Close()
 	}
-	if len(updates.AssetUpdates) > 0 {
+	if len(updates.AssetUpdates) > 0 && len(updates.AssetUpdates[0]) > 0 {
 		any = true
-		// Create new account_asset, or initialize a previously destroyed asset.
+
+		// Create new account_asset, initialize a previously destroyed asset, or apply the balance delta.
 		seta, err := tx.Prepare(`INSERT INTO account_asset (addr, assetid, amount, frozen, created_at) VALUES ($1, $2, $3, $4, $5) ON CONFLICT (addr, assetid) DO UPDATE SET amount = account_asset.amount + EXCLUDED.amount`)
 		if err != nil {
 			return fmt.Errorf("prepare set account_asset, %v", err)
 		}
 		defer seta.Close()
-		for addr, aulist := range updates.AssetUpdates {
-			for _, au := range aulist {
-				if au.AssetID == debugAsset {
-					db.log.Errorf("%d axfer %s %s", round, b64(addr[:]), obs(au))
-				}
-				if au.Delta.IsInt64() {
-					// easy case
-					delta := au.Delta.Int64()
-					// don't skip delta == 0; mark opt-in
-					_, err = seta.Exec(addr[:], au.AssetID, delta, au.DefaultFrozen, round)
-					if err != nil {
-						return fmt.Errorf("update account asset, %v", err)
-					}
-				} else {
-					sign := au.Delta.Sign()
-					var mi big.Int
-					var step int64
-					if sign > 0 {
-						mi.SetInt64(math.MaxInt64)
-						step = math.MaxInt64
-					} else if sign < 0 {
-						mi.SetInt64(math.MinInt64)
-						step = math.MinInt64
-					} else {
-						continue
+
+		// On asset opt-out attach some extra "apply data" metadata to allow rewinding the asset close if requested.
+		acc, err := tx.Prepare(`WITH aaamount AS (SELECT ($1)::bigint as round, ($2)::bigint as intra, x.amount FROM account_asset x WHERE x.addr = $3 AND x.assetid = $4)
+UPDATE txn ut SET extra = jsonb_set(coalesce(ut.extra, '{}'::jsonb), '{aca}', to_jsonb(aaamount.amount)) FROM aaamount WHERE ut.round = aaamount.round AND ut.intra = aaamount.intra`)
+		if err != nil {
+			return fmt.Errorf("prepare asset close0, %v", err)
+		}
+		defer acc.Close()
+		// On asset opt-out update the CloseTo account_asset
+		acs, err := tx.Prepare(`INSERT INTO account_asset (addr, assetid, amount, frozen, created_at)
+SELECT $1, $2, x.amount, $3, $6 FROM account_asset x WHERE x.addr = $4 AND x.assetid = $5
+ON CONFLICT (addr, assetid) DO UPDATE SET amount = account_asset.amount + EXCLUDED.amount`)
+		if err != nil {
+			return fmt.Errorf("prepare asset close1, %v", err)
+		}
+		defer acs.Close()
+		// On asset opt-out mark the account_asset as closed with zero balance.
+		acd, err := tx.Prepare(`UPDATE account_asset SET amount = 0, closed_at = $1 WHERE addr = $2 AND assetid = $3`)
+		if err != nil {
+			return fmt.Errorf("prepare asset close2, %v", err)
+		}
+		defer acd.Close()
+
+
+		for _, subround := range updates.AssetUpdates {
```

Comment:
Idea for refactoring this. Keep the fused update-close apply section of code here, but move the 'subround' behavior into the existing aulist []AssetUpdate. In updateAsset() search the aulist from the end. If the item found has a close (or as is now, if no AU for the assetid), create a new AssetUpdate and append it. There will now be two AssetUpdate records in the list to be applied. So a close & reopen in one round should do the right thing.
This rearranges the complexity; not sure if it's better than the 'subround' way. It might create fewer database Exec statements by allowing all the assets not closed to continue aggregating txns into one update delta for the round.

Reply:
I like this idea and spent a little time trying to get it to work before the current implementation. I wanted to keep as much of the complexity in the accounting package as possible and avoid changing the type of `updates.AssetUpdates`.

The problem occurs when there are multiple assets closing into each other. When any asset closes it might close into another asset which will close later. It is probably possible to do some sort of dependency resolution and recursively commit the things which need to be committed in the right order.

My hunch is that in the vast majority of cases there will only be a single "subround", so this sort of optimization probably isn't worthwhile.

Reply:
Two senses of the ultimate problem here:
* Indexer tried to prematurely optimize expecting that PostgreSQL would be slow, so aggregate transactions in Go code and apply fewer comprehensive transactions that achieve the net delta of the round. This emulation of the is tricky and not quite right.
* Indexer is also not following the algod model of storing all the AccountData in one blob, fetching it, and writing back one new AccountData blob at the end of the round. Feeding actions to PostgreSQL to make the read-modify-write happen server side should be faster, but is just different than what the accounting logic really wants to do. Algorand transactions only sometimes have all the delta they need to apply that way. Payment txns work better that way.

There's been some talk of "maybe we should make the algod accounting logic a library that both algod and Indexer can include". It would be _two_ big projects to make that change on the algod side and then make that change on the Indexer.